### PR TITLE
[SPARK-42070][SQL] Change the default value of argument of Mask function from -1 to NULL

### PR DESCRIPTION
**Pull request by vinodkc**
### What changes were proposed in this pull request?

In the udf 'mask', using -1 as ignore parameter in the position of String type  argument is not a standard way, need to change the value of ignore argument from -1 to NULL.
This PR changes the value of ignore argument in mask udf from -1 to NULL.

### Why are the changes needed?

To standardize the imput value and type

### Does this PR introduce _any_ user-facing change?

Yes, value of ignore params changed from -1 to NULL

### How was this patch tested?

Added Tests

### CODE REVIEW

**Comment by vinodkc** [comment]
@dtenedor , @srielau , Can you please review this PR

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 1]
```diff
@@ -17,27 +17,29 @@
...
     Arguments:
       * input      - string value to mask. Supported types: STRING, VARCHAR, CHAR
-      * upperChar  - character to replace upper-case characters with. Specify -1 to retain original character. Default value: 'X'
-      * lowerChar  - character to replace lower-case characters with. Specify -1 to retain original character. Default value: 'x'
-      * digitChar  - character to replace digit characters with. Specify -1 to retain original character. Default value: 'n'
-      * otherChar  - character to replace all other characters with. Specify -1 to retain original character. Default value: -1
+      * upperChar  - character to replace upper-case characters with. Specify NULL to retain original character. Default value: 'X'
+      * lowerChar  - character to replace lower-case characters with. Specify NULL to retain original character. Default value: 'x'
+      * digitChar  - character to replace digit characters with. Specify NULL to retain original character. Default value: 'n'
+      * otherChar  - character to replace all other characters with. Specify NULL to retain original character. Default value: NULL
```
Comment (srielau) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:42]:
```suggestion
      * otherChar  - character to replace all other characters with. Specify NULL to retain original character. Default value: Retain original character.
```

**Review by srielau — APPROVED** [review]

**Comment by srielau** [comment]
> @dtenedor , @srielau , Can you please review this PR

Thanks!

**Review by srowen — COMMENTED** [review]
This would be a breaking change though, right? not sure that is feasible

**Comment by dtenedor** [comment]
> This would be a breaking change though, right? not sure that is feasible

@srowen technically this true. However, the previous PR to implement the `mask` function was merged relatively recently and no Spark release has occurred since then, IIUC. IMO, it seems OK to make this change in the interest of conforming to ANSI SQL. 👍

**Review by dtenedor — APPROVED** [review]

**Comment by srowen** [comment]
OK, this was in https://issues.apache.org/jira/browse/SPARK-40686 - worth mentioning the connection in the JIRA maybe.
CC @gengliangwang

**Comment by vinodkc** [comment]
@srowen , I updated  both [SPARK-42070](https://issues.apache.org/jira/browse/SPARK-42070) and [SPARK-40687](https://issues.apache.org/jira/browse/SPARK-40687) with the proposed change and reason

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 2]
```diff
@@ -81,34 +83,38 @@ case class Mask(
     otherChar: Expression)
     extends QuinaryExpression
     with ImplicitCastInputTypes
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:85]: Do we still need this `ImplicitCastInputTypes` trait?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 3]
```diff
@@ -207,6 +204,37 @@ case class Mask(
...
 
+  /**
+   * Short hand for generating quinary evaluation code. If either of the sub-expressions is null,
+   * the result of this computation is assumed to be null.
+   *
+   * @param f
+   *   function that accepts the 5 non-null evaluation result names of children and returns Java
+   *   code to compute the output.
+   */
+  override protected def nullSafeCodeGen(
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:215]: Shall we put this into `QuinaryExpression` instead?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala (hunk 2 above)
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:85]: yes, still overriding `def inputTypes: Seq[AbstractDataType]` from ImplicitCastInputTypes <- ExpectsInputTypes

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:85]: I think we can expect all the inputs to be string-type. `ExpectsInputTypes` should be enough here.

Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala]: Done, changed to ExpectsInputTypes

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala [hunk 4]
```diff
@@ -17,27 +17,29 @@
...
     Arguments:
       * input      - string value to mask. Supported types: STRING, VARCHAR, CHAR
-      * upperChar  - character to replace upper-case characters with. Specify -1 to retain original character. Default value: 'X'
-      * lowerChar  - character to replace lower-case characters with. Specify -1 to retain original character. Default value: 'x'
-      * digitChar  - character to replace digit characters with. Specify -1 to retain original character. Default value: 'n'
-      * otherChar  - character to replace all other characters with. Specify -1 to retain original character. Default value: -1
+      * upperChar  - character to replace upper-case characters with. Specify NULL to retain original character. Default value: 'X'
+      * lowerChar  - character to replace lower-case characters with. Specify NULL to retain original character. Default value: 'x'
+      * digitChar  - character to replace digit characters with. Specify NULL to retain original character. Default value: 'n'
+      * otherChar  - character to replace all other characters with. Specify NULL to retain original character. Default value: Retain original character
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala:42]: Nit: default value should be NULL?

**Review by gengliangwang — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala (hunk 4 above)
Comment (vinodkc) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/maskExpressions.scala]: updated Default value: NULL

**Comment by gengliangwang** [comment]
@vinodkc thanks for the change. Merging to master.

**Pull request closed without merging**